Proofs of some linearity axioms.

capn-freako opened the pull request:
A definition of "linear map" along w/ some proofs of a few axioms of linearity.

Taneb commented:
Looking at this, I'm not convinced that this is the right design.

* `¬-involutive-≈ᴬ`, `¬-involutive-≈ᴮ`, `-ᴬ‿involutive`, and `-ᴮ‿involutive` are properties of the modules, not of the morphism.
* Furthermore, `-ᴬ‿involutive` and `-ᴮ‿involutive` are already implied by the `Module` bundle. We can create a proof of this by `Algebra.Properties.Group.⁻¹-involutive +ᴹ-group`.
* Removing those fields makes the `LinMap` record just a bundle of a function with an `IsModuleHomomorphism`. I'd personally like to see homomorphism bundles added to the library, but this isn't the way to do it. Without those bundles, I'd pass a function and a proof it's a homomorphism separately into the anonymous module in `Algebra.Module.Morphism.Linear.Properties`
* I suspect `¬-involutive-≈` lets us prove that equality for the elements of the module in general is decidable (I'm not sure about this). If that's true, I think it'd be worth adding algebraic structures with decidable equality as a thing in the library.
* We _really_ shouldn't postulate things like Extensionality in stdlib. If it's necessary we should pass it as a parameter. In this case it doesn't seem to actually be being used, so it should be OK to remove it.

I'm going to open some tickets to discuss designs for what I'd like to see exist

JacquesCarette commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,127 @@
+    Σ[ (s , z) ∈ S × A ] ((s ·ᴬ (x +ᴬ -ᴬ y) ≈ᴬ z) × (f z ≉ᴮ 0ᴮ)) →
+    f x ≈ᴮ f y → x ≈ᴬ y
+  inj-lm {x = x} {y = y} ((s , z) , (s·[x-y]≈z , fz≉𝟘)) fx≈fy =
+    let fx-fy≈𝟘 : f x +ᴮ (-ᴮ f y) ≈ᴮ 0ᴮ
```
nested lets is not as good as nested where -- lets are inlined.

capn-freako commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,127 @@
+    modA modB      : Module ring m ℓm
+  
+postulate
+  extensionality  : Extensionality ℓ₁ ℓ₂
```
Removed; arriving in next push.

Taneb commented:
> I'm new to Agda and, particularly, to its standard library. So, I need to ask a few questions of clarification. Thanks, in advance, for your patience!

No worries, glad to have y
 
>     1. Can you tell me what you mean by "function-and-structure option"?

The `LinMap` record you've defined contains a function and a bunch of properties of that function. Until now, we've defined records that look like `IsFooHomomorphism`, parameterized on the function. It'd fit better with the standard library to have a record `IsLinearMap (f : A -> B)`. and pass the function and the proof that it's a linear map as separate arguments. (In addition, note that `IsLinearMap` should be the same as `IsModuleHomomorphism`)

>     2. Where do you think _algebraic structures w/ decidable equality_ should live?

I think this should be a bigger discussion, honestly. I'm going to make another issue about this. (as an aside, I think I was wrong when I said that your `¬-involutive-≈` properties give the module decidable equality. I've looked around and thought a bit, and it's a weaker condition that's called "stability" elsewhere in the library

MatthewDaggitt commented:
Thank you for the great first PR, we'd definitely welcome the content! I agree with the comments given by @Taneb and @JacquesCarette above though. Looking forward to seeing the updated version :+1:

capn-freako commented:
> The `LinMap` record you've defined contains a function and a bunch of properties of that function. Until now, we've defined records that look like `IsFooHomomorphism`, parameterized on the function. It'd fit better with the standard library to have a record `IsLinearMap (f : A -> B)`. and pass the function and the proof that it's a linear map as separate arguments. (In addition, note that `IsLinearMap` should be the same as `IsModuleHomomorphism`)

Ah, okay. So, it sounds like what you're proposing is:

1. Skip my definition of `LinMap` entirely.
2. Pack my linearity axiom proofs into a module that is predicated upon an `IsModuleHomomorphism` instance.

Or, were you suggesting that I add my proofs to the _existing_ definition of `IsModuleHomomorphism`?
If a new module, should it be a Bundle?

Thanks,
-db

capn-freako commented:
Okay, @Taneb & @MatthewDaggitt , I think I've addressed all your concerns.  
Please, let me know if I've missed something.

Thanks, again, for your helpful guidance!  
-db

JacquesCarette commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,206 @@
+open import Relation.Nullary.Negation using (contraposition)
+
+open MorphismStructures.ModuleMorphisms modA modB
+open Module modA
```
Have you considered doing
```agda
module A = Module modA
module B = Module modB
```
instead, and then using qualified names? agda-categories uses this later style, and it largely quite readable.

Note that I'm not saying that you *have* to do this, merely recommending that you give it an honest try, and see if you like it or not.

capn-freako replied:
Doesn't that collide with my use of "A" and "B" for the carrier types of those modules?

JacquesCarette replied:
Turns out that no, it doesn't: different name spaces for symbols and modules. So you can name them both A and B without problem.

capn-freako replied:
Okay, I switched to your suggested notation.
I think, visually, I prefer the operators w/ superscript "M" replaced by "A"/"B", but I _really_ like getting rid of that long list of renames!

**These changes included in latest push and ready for review.**

JacquesCarette commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,206 @@
+  ; 0#      to 𝟘
+  ; 1#      to 𝟙
+  )
+open module Reasoningᴬ = Reasoning ≈ᴬ-setoid
```
Needing to reason in 2 different setoids is also something that happens a lot. Rather than opening one of them globally (the one for `modB` in this case) and then renaming everything for `modA`, have you considered opening neither of them globally, but use the 'appropriate' one by using `where open ...` in the proofs themselves?

capn-freako replied:
Ah, right, I remember seeing that "trick" in some code and taking a mental note to use it.
Then, of course, I forgot. :(
Let me see where I can apply that, in my own code...

capn-freako replied:
**Changed included in latest push and ready for review.**

JacquesCarette commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,206 @@
+_≉ᴮ_ : B → B → Set ℓm
+x ≉ᴮ y = ¬ (x ≈ᴮ y)
+
+module _
```
One of the things you could do is to open reasoning in `modA` locally to this module, for example.

capn-freako replied:
Okay, let me give that a try.
Thanks for the tip! :)

capn-freako replied:
@JacquesCarette , can I open one or the other for the `module _` in general and then singly override my choice, using a `where open ...` that is local to a particular proof?
When I try this, Agda tells me that I've created ambiguity around the symbol: `begin_`; so, I'm guessing the answer is, "no."

I'm guessing the only alternative is to use a proof-local `where open ...` clause in _every_ proof that uses the equational reasoning symbols?

JacquesCarette replied:
You can't override what's visible in a name space. 

Note if you open things in a `private` block, that visibility won't leak out to the top-level (which is what happens with`_` modules). Sometimes I name my modules, and open them later, to also deal with visibility issues.

`where open ...` in every proof is always an option. But it does get noisy.

capn-freako replied:
Okay, I went with per-proof `where open ...` and it looks fine to my eye.

**These changes included in latest push and ready for review.**

JacquesCarette commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,206 @@
+  open IsModuleHomomorphism isModuleHomomorphism
+
+  -- f(x) ≈ 0 iff x ≈ 0, for linear non-trivial f
+  f𝟘≈𝟘 : {x : A} → ⟦ 0ᴬ ⟧ ≈ᴮ 0ᴮ
```
This is kind of a weird name, as there is no `f` in the rhs.  You should be consistent about what you name the homomorphism. I have no strong preference between `⟦_⟧` and `f`, but strong preference that you make a choice.

I have a weak preference for `f`. `⟦_⟧` is "semantics brackets", and this isn't quite semantics.

JacquesCarette replied:
Also, this is a really weird property! You have an implicit `x` in the assumptions that is not used in the conclusion. It's never ever going to be inferred, so it will always need to be given.

I think you can use `1` instead of `x`. Your proof just needs that `A` is inhabited to go through, and `1` gives you that. You could even use `0` !  But `x` is bad.

capn-freako replied:
D'oh! Yes, I'd intended to change all `f`s to `⟦_⟧`s and missed this one. :(

Oh, _very_ interesting that you actually prefer `f`. I do, as well, as I'd much prefer that `⟦_⟧` only ever meant, "the meaning of." I guess I misread an intended convention when perusing the other code in the standard library.
I'll revert to `f`.

I'll also get rid of the unnecessary `x`.

capn-freako replied:
I believe that I've addressed everything here.

**These changes included in latest push and ready for review.**

capn-freako commented:
A question, regarding Git protocol: Am I supposed to click the _Resolve Conversation_ button when I _think_ I've addressed the concern, or is it for the requestor to click only when they're satisfied that their concern has actually been addressed?

MatthewDaggitt commented:
> A question, regarding Git protocol: Am I supposed to click the Resolve Conversation button when I think I've addressed the concern, or is it for the requestor to click only when they're satisfied that their concern has actually been addressed?

I think it's fine either way. If you have doubts about whether you've addressed properly then I'd leave it open, otherwise feel free to hit "Resolve" :+1:

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,172 @@
+    B.0ᴹ ∎
+    where open Reasoning B.≈ᴹ-setoid
+
+  f-x≈-fx : {x : A} → f (A.-ᴹ x) B.≈ᴹ B.-ᴹ f x
```
This name should be `f[-x]≈-fx`

capn-freako replied:
Changed as requested.

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,172 @@
+           )
+
+  -- f is odd (i.e. - f (-x) ≈ - (f x)).
+  fx+f-x≈𝟘 : {x : A} → f x B.+ᴹ f (A.-ᴹ x) B.≈ᴹ B.0ᴹ
```
Same with this one.

capn-freako replied:
Changed as requested.

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,172 @@
+      f (s A.*ₗ x) ≈⟨ fs·x≈fy ⟩
+      f y ∎
+    s·fx≉𝟘  : (s B.*ₗ f x) ≉ᴮ B.0ᴹ
+    s·fx≉𝟘  = λ s·fx≈𝟘 →
```
No need to have a lambda here, this can go on the LHS of the equals, e.g. `s·fx≉𝟘 s·fx≈𝟘 =`

capn-freako replied:
Changed as requested.

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,172 @@
+import Relation.Binary.Reasoning.Setoid   as Reasoning
+import Algebra.Module.Morphism.Structures as MorphismStructures
+
+open import Agda.Builtin.Sigma
```
We shouldn't import the builtin modules. Everything from here should be re-exported by `Data.Product`.

capn-freako replied:
Changed as requested.

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,172 @@
+open MorphismStructures.ModuleMorphisms modA modB
+module A = Module modA
+open A using () renaming (Carrierᴹ to A)
+open Properties A.leftSemimodule
```
Again instead of the renaming, we can overload the module name `A` and write:
```agda
module A = Properties A.leftSemimodule

.... A.non-zeroʳ ....
```

capn-freako replied:
When I try this, I'm told:

```
Duplicate definition of module A. Previous definition of (record)
module A at
```

capn-freako replied:
Since I've already got modules "A" and "B", I've chosen "PA" and "PB" for the _properties_ of modules A and B, respectively.

```agda
module A = Module modA
open A using () renaming (Carrierᴹ to A)
module B = Module modB
open B using () renaming (Carrierᴹ to B)
module PA = ModuleProperties modA
module PB = ModuleProperties modB
```

MatthewDaggitt reviewed (commented):
You haven't picked the simplest first PR @capn-freako, thank you for sticking it out :smile: 

The module structure now looks pretty good to me, so I've left some comments about the style.

Just a heads up that there'll probably be one more lot of comments after this round.

capn-freako commented on src/Algebra/Module/Bundles.agda:
```diff
@@ -333,3 +334,7 @@ record Module (commutativeRing : CommutativeRing r ℓr) m ℓm
 
   open Semimodule semimodule public using (*ₗ-comm; *ᵣ-comm)
+
+  -ᴹ‿involutive : {x : Carrierᴹ} → -ᴹ (-ᴹ x) ≈ᴹ x
```
Changed as requested.

capn-freako commented on src/Algebra/Module/Bundles.agda:
```diff
@@ -333,3 +334,7 @@ record Module (commutativeRing : CommutativeRing r ℓr) m ℓm
   open Semimodule semimodule public using (*ₗ-comm; *ᵣ-comm)
+
+  -ᴹ‿involutive : {x : Carrierᴹ} → -ᴹ (-ᴹ x) ≈ᴹ x
+  -ᴹ‿involutive {x = x} = ⁻¹-involutive +ᴹ-group x
```
Changed as requested.

capn-freako commented on CHANGELOG.md:
````diff
@@ -950,6 +950,12 @@ New modules
 
+* Proofs of some axioms of linearity:
+  ```
+  Algebra.Module.Bundles
````
Changed as requested.

capn-freako commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,172 @@
+open Function.Definitions A._≈ᴹ_ B._≈ᴹ_
+
+_≉ᴬ_ : A → A → Set ℓm
+x ≉ᴬ y = ¬ (x A.≈ᴹ y)
```
Changed as requested.

capn-freako commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,172 @@
+  (modA modB  : Module ring m ℓm)
+  where
+
+import Algebra.Module.Properties          as Properties
```
Changed as requested.

capn-freako commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,172 @@
+  using (_≈_; _*_; _+_) renaming
+  ( Carrier to S
+  ; 0#      to 𝟘
+  ; 1#      to 𝟙
```
Changed as requested.

capn-freako commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,172 @@
+  ; 0#      to 𝟘
+  ; 1#      to 𝟙
+  )
+open Function.Definitions A._≈ᴹ_ B._≈ᴹ_
```
Correct. Removed.

capn-freako commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,172 @@
+    f A.0ᴹ          ≈⟨ ⟦⟧-cong (A.≈ᴹ-sym (A.*ₗ-zeroˡ A.0ᴹ)) ⟩
+    f (𝟘 A.*ₗ A.0ᴹ) ≈⟨ *ₗ-homo 𝟘 A.0ᴹ ⟩
+    𝟘 B.*ₗ f A.0ᴹ   ≈⟨ B.*ₗ-zeroˡ (f A.0ᴹ) ⟩
+    B.0ᴹ ∎
```
Changed as requested.

capn-freako commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,172 @@
+    where open Reasoning B.≈ᴹ-setoid
+
+  x≈𝟘→fx≈𝟘 : {x : A} → x A.≈ᴹ A.0ᴹ → f x B.≈ᴹ B.0ᴹ
+  x≈𝟘→fx≈𝟘 {x = x} x≈𝟘 = begin
```
Changed as requested.

capn-freako commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,172 @@
+      ¬x≉𝟘 = λ x≉𝟘 → zero-unique ((s , y) , (s·x≈y , fy≉𝟘)) x≉𝟘 fx≈𝟘
+
+    -- A non-trivial linear function is injective.
+    fx-fy≈𝟘 : {x y : A} {fx≈fy : f x B.≈ᴹ f y} → f x B.+ᴹ (B.-ᴹ f y) B.≈ᴹ B.0ᴹ
```
Thanks! Your comment _really_ helped me gain some additional clarity on implicit arguments, telescopes, etc.

Changed as requested.

capn-freako commented on src/Algebra/Module/Properties.agda:
```diff
@@ -0,0 +1,57 @@
+  renaming
+  ( Carrierᴹ to M
+  ; _*ₗ_     to _·_
+  ; 0ᴹ       to 𝟘
```
Changed as requested.

capn-freako commented on src/Algebra/Module/Properties.agda:
```diff
@@ -0,0 +1,57 @@
+
+non-zeroˡ : {s : R} {v : M} → s · v ≉ᴹ 𝟘 → s ≉ 0ᴿ
+non-zeroˡ {s = s} {v = v} s·v≉𝟘 = λ { s≈0ᴿ →
+  let s·v≈𝟘 : s  · v ≈ᴹ 𝟘
```
Is this what you had in mind?

```agda
s≈𝟘⇒s*v≈𝟘 : ∀ {s} {v} → s ≈ 0# → s *ₗ v ≈ᴹ 0ᴹ
s≈𝟘⇒s*v≈𝟘 {s} {v} s≈𝟘 = begin
  s  *ₗ v ≈⟨ *ₗ-congʳ s≈𝟘 ⟩
  0# *ₗ v ≈⟨ *ₗ-zeroˡ v ⟩
  0ᴹ      ∎

v≈𝟘⇒s*v≈𝟘 : ∀ {s} {v} → v ≈ᴹ 0ᴹ → s *ₗ v ≈ᴹ 0ᴹ
v≈𝟘⇒s*v≈𝟘 {s} {v} v≈𝟘 = begin
  s *ₗ v  ≈⟨ *ₗ-congˡ v≈𝟘 ⟩
  s *ₗ 0ᴹ ≈⟨ *ₗ-zeroʳ s ⟩
  0ᴹ      ∎

non-zeroˡ : ∀ {s} {v} → s *ₗ v ≉ᴹ 0ᴹ → s ≉ 0#
non-zeroˡ = contraposition s≈𝟘⇒s*v≈𝟘

non-zeroʳ : ∀ {s} {v} → s *ₗ v ≉ᴹ 0ᴹ → v ≉ᴹ 0ᴹ
non-zeroʳ = contraposition v≈𝟘⇒s*v≈𝟘
```

capn-freako replied:
@MatthewDaggitt ?

capn-freako commented on src/Algebra/Module/Properties.agda:
```diff
@@ -0,0 +1,57 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Properties of scaling.
```
Changed as requested.

MatthewDaggitt commented on src/Algebra/Module/Bundles.agda:
```diff
@@ -92,6 +99,10 @@ record LeftModule (ring : Ring r ℓr) m ℓm : Set (r ⊔ ℓr ⊔ suc (m ⊔ 
 
+  infix 4 _≉ᴹ_
+  _≉ᴹ_ : Rel Carrierᴹ _
+  a ≉ᴹ b = ¬ (a ≈ᴹ b)
```
We shouldn't have to redefine this more than twice. For example here, we can import it from `leftSemimodule` below right?

capn-freako replied:
D'oh! I know better than to be unnecessarily redundant w/ code; sorry. :(
Fixed and arriving w/ next push.

MatthewDaggitt commented on src/Algebra/Module/Bundles.agda:
```diff
@@ -333,3 +368,6 @@ record Module (commutativeRing : CommutativeRing r ℓr) m ℓm
 
   open Semimodule semimodule public using (*ₗ-comm; *ᵣ-comm)
+
+  -ᴹ-involutive : Involutive _≈ᴹ_ -ᴹ_
```
The bundle (i.e. `Semimodule`) should not contain any proofs that the structures (i.e. `IsSemimodule`) does not have. Therefore this should go in `Algebra.Module.Properties.Semimodule` (this file may not yet exist).

capn-freako replied:
You meant to say `Module`, right?
(This property is defined inside the `Module` record and relies on `-ᴹ_`.)

capn-freako replied:
So, I just noticed this in `Algebra.Module.Bundles.Semimodule`:

```agda
*ₗ-comm : L.Commutative _*ₗ_
*ₗ-comm x y m = begin
  x *ₗ y *ₗ m   ≈⟨ ≈ᴹ-sym (*ₗ-assoc x y m) ⟩
  (x * y) *ₗ m  ≈⟨ *ₗ-cong (*-comm _ _) ≈ᴹ-refl ⟩
  (y * x) *ₗ m  ≈⟨ *ₗ-assoc y x m ⟩
  y *ₗ x *ₗ m   ∎

*ᵣ-comm : R.Commutative _*ᵣ_
*ᵣ-comm m x y = begin
  m *ᵣ x *ᵣ y   ≈⟨ *ᵣ-assoc m x y ⟩
  m *ᵣ (x * y)  ≈⟨ *ᵣ-cong ≈ᴹ-refl (*-comm _ _) ⟩
  m *ᵣ (y * x)  ≈⟨ ≈ᴹ-sym (*ᵣ-assoc m y x) ⟩
  m *ᵣ y *ᵣ x   ∎
```

And I'm wondering why this exception was made.
Thanks!

capn-freako replied:
Here's what I ended up doing:

```
dbanas@DESKTOP-G84ND7C:~/prj/agda-stdlib
$ tree src/Algebra/Module/
src/Algebra/Module/
{snip}
├── Properties
│   ├── Semimodule.agda - Contains the old contents of `../Properties.agda`.
├── Properties.agda     - Provides `-ᴹ-involutive` and
                           imports (publicly) `Properties/Semimodule.agda.
{snip}
```

Let me know if you'd like further changes.

capn-freako replied:
@MatthewDaggitt ?

MatthewDaggitt commented on src/Algebra/Module/Properties.agda:
```diff
@@ -0,0 +1,47 @@
+open LeftSemimodule lsmod
+  using ( *ₗ-zeroˡ; *ₗ-zeroʳ; ≈ᴹ-setoid; *ₗ-congˡ; *ₗ-congʳ
+        ; _*ₗ_; 0ᴹ; _≈ᴹ_; _≉ᴹ_)
+  renaming (Carrierᴹ to M)
```
No longer need this renaming and the renaming below. So we can probably just `open LeftSemimodule` without the `using`?

capn-freako replied:
Yep, thanks!
Fixed and arriving in next push.

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,163 @@
+
+open import Algebra                   using (CommutativeRing)
+open import Algebra.Module            using (Module)
+open import Level                     using (Level; _⊔_)
```
Don't need `_⊔_`?

capn-freako replied:
Fixed and arriving in next push.

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,163 @@
+  where
+
+import Algebra.Module.Properties          as ModuleProperties
+import Function.Definitions
```
Don't need this import?

capn-freako replied:
Nope; removed in next push.

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,163 @@
+--   using () renaming
+--   ( non-zeroʳ to non-zeroʳᴬ
+--   ; non-zeroˡ to non-zeroˡᴬ
+--   )
```
Please remove the commented out code :smile:

capn-freako replied:
Removed in next push.

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,163 @@
+  open IsModuleHomomorphism isModuleHomomorphism
+
+  -- f(x) ≈ 0 iff x ≈ 0, for linear non-trivial f
+  f𝟘≈𝟘 : f A.0ᴹ B.≈ᴹ B.0ᴹ
```
Please could we use standard `0`s in the names rather than boldfaced ones. More standard, and easier to type!

capn-freako replied:
Changed as requested in next push.

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,163 @@
+    f (0# A.*ₗ A.0ᴹ) ≈⟨ *ₗ-homo 0# A.0ᴹ ⟩
+    0# B.*ₗ f A.0ᴹ   ≈⟨ B.*ₗ-zeroˡ (f A.0ᴹ) ⟩
+    B.0ᴹ             ∎
+    where open Reasoning B.≈ᴹ-setoid
```
As an alternative to reopening `Reasoning.Setoid` in every proof, we could use `Relation.Binary.Reasoning.MultiSetoid`?

capn-freako replied:
Oh, that is awesome sauce!!! Thanks. :)

Changed as requested in next push.

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,163 @@
+      B.0ᴹ                ∎
+      where open Reasoning B.≈ᴹ-setoid
+
+    fx-y≈𝟘 : ∀ {x} {y} → (f x B.≈ᴹ f y) → f (x A.+ᴹ (A.-ᴹ y)) B.≈ᴹ B.0ᴹ
```
Can combine `{x} {y}` into `{x y}` and no need for brackets around `(f x B.≈ᴹ f y)`

capn-freako replied:
Cleaned up as requested in next push.

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,163 @@
+  f[-x]≈-fx : ∀ {x} → f (A.-ᴹ x) B.≈ᴹ B.-ᴹ f x
+  f[-x]≈-fx {x} = B.uniqueʳ‿-ᴹ (f x) (f (A.-ᴹ x)) fx+f[-x]≈𝟘
+
+  module _ {dne : DoubleNegationElimination _} where
```
Hmm assuming double negation elimination (for every type!) is very strong here! First of all is definitely needed for all of these? (just checking!)

Secondly can we get away with assuming the (weaker?) statement of decidability of `A.≉ᴹ`?

capn-freako replied:
No, I only need: `¬x≉0⇒x≈0`. I just wasn't sure how else to get that.

You think I can get it, via decidability?
Can you give me a bit more of a hint?

JacquesCarette replied:
I would put that locally as an assumption only where you need it.  Yes, this is derivable from decidability.

capn-freako replied:
> I would put that locally as an assumption only where you need it.

Yeah, that makes sense since I only use it for one lemma.

**==> Requested change arriving in next push.**

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,163 @@
+  module _ {dne : DoubleNegationElimination _} where
+
+    fx≈𝟘⇒x≈𝟘 : ∀ {x} →
+      Σ[ (s , y) ∈ S × A ] ((s A.*ₗ x A.≈ᴹ y) × (f y B.≉ᴹ B.0ᴹ)) →
```
It would be much shorter to write: `∃₂ λ s y →` rather than `Σ[ (s , y) ∈ S × A ]` in this proof and the one below.

capn-freako replied:
Hmmm, I can't seem to make that work.
When I try: `∃₂ λ s y →`, I'm told:

```
Cannot eliminate type
Σ (CommutativeRing.Carrier ring₁)
(λ a →
   ∃
   (λ y →
      a A.*ₗ x A.≈ᴹ y × f y B.≉ᴹ Module.0ᴹ modB →
      (modB Module.≈ᴹ f x) (Module.0ᴹ modB) → x A.≈ᴹ A.0ᴹ))
with constructor pattern (s , y) , s·x≈y , fy≉0 (did you supply too
many arguments?)
```

And when I try: `∃ λ (s , y) →`, I'm told:

```
Cannot eliminate type
Σ
(_A_256 (modA = modA) (modB = modB)
 (isModuleHomomorphism = isModuleHomomorphism))
(_258 (modA = modA) (modB = modB)
 (isModuleHomomorphism = isModuleHomomorphism))
with constructor pattern (s , y) , s·x≈y , fy≉0 (did you supply too
many arguments?)
```

capn-freako replied:
@MatthewDaggitt ?

MatthewDaggitt replied:
You will need to put brackets around it, e.g. `(∃₂ λ s y → ((s A.*ₗ x A.≈ᴹ y) × (f y B.≉ᴹ B.0ᴹ)))`. The argument can then be `(s , y , s·x≈y , fy≉0)`.

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,163 @@
+    y≉𝟘     : y A.≉ᴹ A.0ᴹ
+    y≉𝟘     = fx≉𝟘⇒x≉𝟘 fy≉𝟘
+    fs·x≈fy : f (s A.*ₗ x) B.≈ᴹ f y
+    fs·x≈fy = ⟦⟧-cong s·x≈y
```
We can just fold the definition of `fs·x≈fy` into the proof below no? It's really short...

capn-freako replied:
Yep, thanks.
Fixed in next push.

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,163 @@
+    fs·x≈fy : f (s A.*ₗ x) B.≈ᴹ f y
+    fs·x≈fy = ⟦⟧-cong s·x≈y
+    s·fx≈fy : s B.*ₗ f x B.≈ᴹ f y
+    s·fx≈fy = begin
```
Hmm this proof doesn't seem to be used at all?

capn-freako replied:
D'oh! Thanks. :(

Removed in next push.

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,163 @@
+  fx≉𝟘⇒x≉𝟘 = contraposition x≈𝟘⇒fx≈𝟘
+
+  -- Zero is a unique output of linear map ≉ `const 𝟘`.
+  zero-unique : ∀ {x} →
```
Might be useful to have a read of the [style guide ](https://github.com/agda/agda-stdlib/blob/master/notes/style-guide.md), especially the bit about indentation.

MatthewDaggitt commented on src/Algebra/Module/Properties.agda:
```diff
@@ -0,0 +1,47 @@
+  s *ₗ 0ᴹ ≈⟨ *ₗ-zeroʳ s ⟩
+  0ᴹ      ∎
+
+non-zeroˡ : ∀ {s} {v} → s *ₗ v ≉ᴹ 0ᴹ → s ≉ 0#
```
This name isn't very descriptive of what the proof says. Maybe we should use the names `s*v≉0⇒s≉0` to match the names of the proofs of the opposite direction above?

capn-freako replied:
Changed as requested in most recent push.

MatthewDaggitt reviewed (commented):
Thanks! Here's another round of comments!

JacquesCarette commented on src/Algebra/Module/Properties/Semimodule.agda:
```diff
@@ -0,0 +1,40 @@
+open import Relation.Nullary.Negation using (contraposition)
+open import Relation.Binary.Reasoning.Setoid ≈ᴹ-setoid
+
+s≈0⇒s*v≈0 : ∀ {s} {v} → s ≈ 0# → s *ₗ v ≈ᴹ 0ᴹ
```
This is true for any left zero (and the next, for any right zero). I get the feeling that it might already be proved in greater generality elsewhere? (Or should be?)

capn-freako replied:
Hi @JacquesCarette ,

I'm guessing that you'd like to see these properties moved into `Algebra.Module.Structures.IsLeftSemimodule`.
Is that correct?
It makes sense, since that's where the last dependencies (i.e. - `*ₗ-congˡ` and `*ₗ-congʳ`) are defined.

However, when I try to move this code into that record definition, I bump into all kinds of awkwardness, which I believe stems from the fact that `*ₗ` is _not_ defined as an infix operator:

```agda
record IsLeftSemimodule (*ₗ : Opₗ R M) : Set (r ⊔ m ⊔ ℓr ⊔ ℓm) where
```

Is this movement of code super important to you?

capn-freako replied:
But, doesn't it need to apply, at least, to a semimodule?
That is, isn't _left zero_ a (semi)module-centric conept?

JacquesCarette commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,135 @@
+
+  open IsModuleHomomorphism isModuleHomomorphism
+
+  -- f(x) ≈ 0 iff x ≈ 0, for linear non-trivial f
```
Now that I relook at this proof: this seems to only assume left-zero in both A and B.

capn-freako replied:
D'oh! You're right. Thank you!

**==> Requested change arriving with next push.**

JacquesCarette commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,135 @@
+  fx≉0⇒x≉0 = contraposition x≈0⇒fx≈0
+
+  -- Zero is a unique output of linear map ≉ `const 0`.
+  zero-unique : ∀ {x} →
```
This is an awkward way to phrase uniqueness (constructively). Normally what you do is assume you have something else (call it `z`) that behaves like a zero, and then derive that it must be equal to 0.

capn-freako replied:
Hi @JacquesCarette ,

Are you proposing a change to:

1. the type signature,
2. the definition, or
3. both?

JacquesCarette replied:
I think both. I'd expect something more positive, in the sense of "assume this thing z behaves just like 0, then it must in fact be equal to 0".  That's what 'uniqueness' brings to mind for me. You've got some kind of double-negative version of that.

capn-freako replied:
Oh, except that I'm going the other way in this lemma.
I'm trying to prove that if the argument to `f` is _not_ zero then the result of `f` cannot be zero:

```agda
x A.≉ᴹ A.0ᴹ → f x B.≉ᴹ B.0ᴹ
```

Perhaps, my function name is misleading?

JacquesCarette commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,135 @@
+             B.0ᴹ         ∎
+           )
+
+  -- f is odd (i.e. - f (-x) ≈ - (f x)).
```
Can't you prove the statement that's in the comment directly? (By linearity, you should be able to pull -1 out. You'd need to first prove that -x is necessarily -1*x, which is useful in its own right).

capn-freako replied:
So, the only thing that occurs to me is to show that `(-1 *)` also forms the additive inverse; something like:

```agda
x+[-x]≈x+[-1*x] : ∀ {x} → x A.+ᴹ (A.-ᴹ x) ≈ x A.+ᴹ (- 1#) A.*ₗ x
x+[-x]≈x+[-1*x] = begin⟨ A.≈ᴹ-setoid ⟩
  x A.+ᴹ (A.-ᴹ x) ≈⟨⟩
  A.0ᴹ ≈⟨⟩
  0# A.*ₗ x ≈⟨⟩
  (1# + (- 1#)) A.*ₗ x ≈⟨⟩
  1# A.*ₗ x + (- 1#) A.*ₗ x ≈⟨⟩
  x A.+ᴹ (- 1#) A.*ₗ x
```

But, there are two problems:

1. The code  above is longer than my existing prerequisite: `fx+f[-x]≈0`, and

2. `_+ᴹ_` isn't guaranteed to be injective, or at least I can't find a definition in the library:

    ```agda
    $ git grep -- 'Injective .*_+ᴹ_' src/Algebra/Module/                                 
    
    $
    ```

JacquesCarette commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,135 @@
+      ¬x≉0 = λ x≉0 → zero-unique ((s , y) , (s·x≈y , fy≉0)) x≉0 fx≈0
+
+    -- A non-trivial linear function is injective.
+    fx-fy≈0 : ∀ {x y} → f x B.≈ᴹ f y → f x B.+ᴹ (B.-ᴹ f y) B.≈ᴹ B.0ᴹ
```
This is not really injectivity, I certainly got confused by the comment.

capn-freako replied:
The comment refers to `inj-lm`.
All lemmas in between the comment and `inj-lm` are prerequisites of `inj-lm`.

capn-freako commented:
Hi @Taneb , @MatthewDaggitt , @JacquesCarette ,

While working on my next PR (support of abstract vector spaces), I restructured my linearity stuff in a way that I _think_ might be a better fit to the intended use model of the Agda standard library.

I didn't want to back-merge those changes to this PR w/o giving you an opportunity to review them first.
You can find this "restructuring" on the `vector-space` branch of my fork:
https://github.com/capn-freako/agda-stdlib/tree/vector-space

The changes are basically:

1. Created `Algebra.Module.Morphism.Bundles` as a home for a new `LinearMap` record.
2. Moved `Algebra.Module.Morphism.Linear.Properties` up a level to: `Algebra.Module.Morphism.Properties`, wrapped in a sub-record: `LinearMapProperties`.

Looking forward to hearing your feedback on this restructuring of my linearity axioms,
-db

JacquesCarette commented:
I find the restructuring fine.

MatthewDaggitt commented:
Sorry I'm in the middle of a month long holiday at the moment. Would you
mind if I had a look when I get back? If Jacques says it looks good, then
it should be fine to add it to the main PR!

On Wed, 15 Jun 2022, 20:26 Jacques Carette, ***@***.***>
wrote:

> I find the restructuring fine.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/agda/agda-stdlib/pull/1767#issuecomment-1156793391>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAP26EQ63EEO26LTZNM7G6TVPIN6PANCNFSM5VMJXYEA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

capn-freako commented:
Okay, great! I'll go ahead and merge the changes here.

And sorry to bother you on vacation.
I hope your having a wonderful time. :)

capn-freako commented:
Hmmm, on second thought, I think I'll wait.
Merging would pull in a lot of extraneous stuff not directly pertinent to these linearity proofs.

I'll get the `vector-space` pull request submitted, so it's ready to go when this one gets accepted.
That should minimize any exposure to the changing structure.

capn-freako commented:
Hi all,

A friend suggested that the large number of commits coming in through this PR may be an impediment to its acceptance.
Would you all prefer that I squash those?
(I'm somewhere in the Journeyman phase of my Git learning and never quite sure when it's appropriate to squash commits.)

Thanks,
-db

capn-freako commented:
Hi all,

I'm wondering what I can do, to bring this to closure.
It's kind of a pain, having my `master` branch blocked by this.
(I know, my bad; should've used a branch; I think I've learned my lesson. ;-) )

Also, I've got a third PR waiting in the wings. (It's dependent upon both this one and my second.)
It completes a "trifecta" that I agreed to do, in this working group I'm participating in, which will provide a _denotation_ for `Data.Vec` in terms of linear maps between abstract vector spaces.
I've been told by the group leader that it will be quite useful towards our common aims.
And I'd like to get it available to the rest, without asking them to work from my fork of the standard library.

Anything I can do, to help close this?

Thanks,
-db
:)

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,127 @@
+    PB.s*v≉0⇒v≉0 s·fx≉0
+    where
+    y≉0     : y A.≉ᴹ A.0ᴹ
+    y≉0     = fx≉0⇒x≉0 fy≉0
```
`y≉0` not necessary?

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,127 @@
+  f[-x]≈-fx {x} = B.uniqueʳ‿-ᴹ (f x) (f (A.-ᴹ x)) fx+f[-x]≈0
+
+  -- A non-trivial linear function is injective.
+  fx-fy≈0 : ∀ {x y} → f x B.≈ᴹ f y → f x B.+ᴹ (B.-ᴹ f y) B.≈ᴹ B.0ᴹ
```
This name doesn't seem great, as it omits a very important pre-condition. How about `fx≈fy⇒fx-fy≈0`?

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,127 @@
+    f x B.+ᴹ (B.-ᴹ f x) ≈⟨ B.-ᴹ‿inverseʳ (f x) ⟩
+    B.0ᴹ                ∎
+
+  fx-y≈0 : ∀ {x y} → f x B.≈ᴹ f y → f (x A.+ᴹ (A.-ᴹ y)) B.≈ᴹ B.0ᴹ
```
Likewise I think `fx≈fy⇒f[x-y]≈0` would be better?

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,127 @@
+      ¬x≉0 = λ x≉0 → zero-unique ((s , y) , (s·x≈y , fy≉0)) x≉0 fx≈0
+
+    inj-lm : ∀ {x y} →
+      Σ[ (s , z) ∈ S × A ] ( (s A.*ₗ (x A.+ᴹ A.-ᴹ y) A.≈ᴹ z)
```
Could you ensure the alignment of this type and others in the file match the [style guide](https://github.com/agda/agda-stdlib/blob/master/notes/style-guide.md#indentation)?

MatthewDaggitt commented on src/Algebra/Module/Properties/Semimodule.agda:
```diff
@@ -0,0 +1,40 @@
+  0# *ₗ v ≈⟨ *ₗ-zeroˡ v ⟩
+  0ᴹ      ∎
+
+v≈0⇒s*v≈0 : ∀ {s} {v} → v ≈ᴹ 0ᴹ → s *ₗ v ≈ᴹ 0ᴹ
```
Incredibly minor but we can save two characters in each type by merging the implicit brackets, e.g. `{s v}`.

Also is there a reason that you've switched from using the names `x` and `y` for elements of the main carrier set in the other module to `v` here?

capn-freako replied:
Changed as requested in next push.

The `s` and `v` are intended to suggest: _scalar_ and _vector_.

MatthewDaggitt replied:
This wasn't changed in the next push. I think consistency is the best, so could we change both the names of the variables in the type and the name of the lemma?

capn-freako replied:
Code changed, as requested:

- 's' => 'x'
- 'v' => 'y'

in next push.

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,127 @@
+    f A.0ᴹ              ≈⟨ 0ᴹ-homo ⟩
+    B.0ᴹ                ∎
+
+  f[-x]≈-fx : ∀ {x} → f (A.-ᴹ x) B.≈ᴹ B.-ᴹ f x
```
I suspect the argument `x` should be explicit rather than implicit here, as there's no subsequent argument that Agda can reliably infer it from. The current setup will work sometimes, but not always.

capn-freako replied:
Ah, got it; thanks!

**==> Code changed, as requested, in next push.**

**Question:** Should `x` and `y` in the following two functions be made explicit, as well?

MatthewDaggitt replied:
I think they should be fine as they're provided by an argument :+1:  Our definition of `Injective` in the library has the same form and we don't have any inference problems with that.

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,127 @@
+open import Algebra.Module            using (Module)
+open import Level                     using (Level)
+
+module Algebra.Module.Morphism.Linear.Properties
```
Finally by library convention, this should be called `Algebra.Module.Morphism.ModuleHomomorphism`. I know it has special significance as linear maps, but I think it is better to put comments to that effect both on the main record definition, and in the header of this module?

MatthewDaggitt reviewed (commented):
> A friend suggested that the large number of commits coming in through this PR may be an impediment to its acceptance.
Would you all prefer that I squash those?
(I'm somewhere in the Journeyman phase of my Git learning and never quite sure when it's appropriate to squash commits.)

Your friend is right in that the commit history is messy and we don't want to add it to the main branch, but don't worry, we will squash all the commits down automatically when it's merged in!

> Anything I can do, to help close this?

Apologies for the delay, I'm back from travelling and will try and be more prompt to replying to it now. Once the latest round of comments are addressed should be pretty close to being ready to merge in :+1: Thanks for your patience!

capn-freako reviewed (commented):
Completed my responses to Matthew's comments.
Will push code changes shortly.

MatthewDaggitt commented on src/Algebra/Module/Properties/Semimodule.agda:
```diff
@@ -0,0 +1,40 @@
+
+open CommutativeSemiring semiring
+open Semimodule          semimod
+open import Relation.Nullary.Negation using (contraposition)
```
This should live in the imports above the module declaration as it doesn't depend on the parameters?

capn-freako replied:
Oh, I thought your previous comment in this regard was referring only to `import ...` lines and _not_ to `open import ...` lines. Now, I'm very confused, because my general sense of the standard library code is that the only `open import ...` lines that get pulled _above_ the main module declaration are those necessary, to specify the types of the main module parameters. Is that not the case?

capn-freako replied:
From the style guide:

> All imports should be placed in a list at the top of the file immediately after the module declaration.
>
> The list of imports should be declared in alphabetical order.
>
>If the module takes parameters that require imports from other files, then those imports only may be placed above the module declaration, e.g.

```agda
open import Algebra using (Ring)

module Algebra.Properties.Ring {a l} (ring : Ring a l) where

      ... other imports
```

MatthewDaggitt replied:
You're completely right! I was getting confused, apologies. Thanks for the spot :smile:

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,124 @@
+open import Data.Product
+open import Relation.Binary.Reasoning.MultiSetoid
+open import Relation.Nullary          using (¬_)
+open import Relation.Nullary.Negation using (contraposition)
```
All imports that don't depend on the module parameters should live above the main module declaration.

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,124 @@
+  NonTrivial x = (∃₂ λ s y → ((s A.*ₗ x A.≈ᴹ y) × (f y B.≉ᴹ B.0ᴹ)))
+  -- NonTrivial x = Σ[ (s , y) ∈ S × A ] ( (s A.*ₗ x A.≈ᴹ y)
+  --                                     × (f y B.≉ᴹ B.0ᴹ)
+  --                                     )
```
Remove commented out definition?

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,124 @@
+
+  open IsModuleHomomorphism isModuleHomomorphism
+
+  NonTrivial : A → Set (r Level.⊔ m Level.⊔ ℓm)
```
A comment above this definition explaining the intuition about why this represents non-triviality would be welcome! Also we should probably make it `private` for the moment....

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,124 @@
+  fx≉0⇒x≉0 = contraposition x≈0⇒fx≈0
+
+  -- Zero is a unique output of non-trivial (i.e. - ≉ `const 0`) linear map.
+  zero-unique : ∀ {x} → NonTrivial x → x A.≉ᴹ A.0ᴹ → f x B.≉ᴹ B.0ᴹ
```
I'm not sure if the current name is very representative of what the lemma actually says. Maybe we should rename this as `x≉0⇒fx≉0`?

MatthewDaggitt commented on src/Algebra/Module/Morphism/Linear/Properties.agda:
```diff
@@ -0,0 +1,127 @@
+
+module _
+  {f : A → B}
+  (isModuleHomomorphism : IsModuleHomomorphism f)
```
So it's important that each merged in PR stands on its own two legs, in case hypothetical latter PRs don't get merged in.
So in the interests of merging in _this_ PR, again could we make this change now and then we can change it if required in #1781?

MatthewDaggitt replied:
Ah so you can either write:
```agda
module Algebra.Module.Morphism.ModuleHomomorphism
  {r ℓr m ℓm : Level}
  {ring      : CommutativeRing r ℓr}
  (modA modB  : Module ring m ℓm)
  {f : Module.Carrierᴹ modA → Module.Carrierᴹ modB}
  (isModuleHomomorphism : IsModuleHomomorphism f)
  where
```
or you can do the following neat trick:
```agda
module Algebra.Module.Morphism.ModuleHomomorphism
  {r ℓr m ℓm : Level}
  {ring      : CommutativeRing r ℓr}
  (modA modB  : Module ring m ℓm)
  (open Module modA using () renaming (Carrierᴹ to A))
  (open Module modB using () renaming (Carrierᴹ to B))
  {f : A → B}
  (isModuleHomomorphism : IsModuleHomomorphism f)
  where
```

capn-freako replied:
Code changed, as requested, in next push.

MatthewDaggitt commented:
Okay so once the module parameters change is made, this is ready :+1:

capn-freako commented:
Thanks, Matthew!

Changes have been pushed.

MatthewDaggitt commented:
Merged in. Thanks for the great PR @capn-freako, and for your patience!

capn-freako commented:
Thanks, Matthew!

And thank you, all, for your helpful guidance and patience during my first attempt to contribute. :)

JacquesCarette commented:
Luckily, this is so new that I don't think anythign relies on it yet. So fixing it shouldn't be a big deal, as there's nothing to deprecate!

JacquesCarette commented:
(or am I wrong and this made it out in 1.7?)

Taneb commented:
@JacquesCarette you were correct, this is not in any released branch

JacquesCarette commented:
That helps a lot - fix away! I agree with all the suggestions you made.

MatthewDaggitt commented:
Yup, @Taneb if you have a PR going to fix these then great. Otherwise can you open an issue under the `v2.0` milestone with the above comments in so that they don't get missed?
